app/com/linkedin/drelephant/mapreduce/MapReduceFetcherHadoop2.java

```diff
@@ -54,36 +53,60 @@
   private static final Logger logger = Logger.getLogger(ElephantFetcher.class);
   private static final int MAX_SAMPLE_SIZE = 200;
   private static final String SAMPLING_ENABLED = "sampling_enabled";
-  // We provide one minute job fetch delay due to the job sending lag from AM/NM to JobHistoryServer HDFS
 
   private URLFactory _urlFactory;
   private JSONFactory _jsonFactory;
   private String _jhistoryWebAddr;
   private FetcherConfigurationData _fetcherConfigurationData;
+  private Cluster _cluster;
+  private Configuration _configuration;
 
   public MapReduceFetcherHadoop2(FetcherConfigurationData fetcherConfData) throws IOException {
     this._fetcherConfigurationData = fetcherConfData;
+    this._configuration = new Configuration();
+    this._jsonFactory = new JSONFactory();
+    initCluster();
+  }
+
+  private void initCluster() throws IOException {
+    this._cluster = new Cluster(_configuration);
 
-    final String jhistoryAddr = new Configuration().get("mapreduce.jobhistory.webapp.address");
+    final String jhistoryAddr = _configuration.get("mapreduce.jobhistory.webapp.address");
+    _jhistoryWebAddr = "http://" + jhistoryAddr + "/jobhistory/job/";
 
     logger.info("Connecting to the job history server at " + jhistoryAddr + "...");
     _urlFactory = new URLFactory(jhistoryAddr);
     logger.info("Connection success.");
-
-    _jsonFactory = new JSONFactory();
-    _jhistoryWebAddr = "http://" + jhistoryAddr + "/jobhistory/job/";
   }
 
   @Override
-  public MapReduceApplicationData fetchData(AnalyticJob analyticJob) throws IOException, AuthenticationException {
+  public MapReduceApplicationData fetchData(AnalyticJob analyticJob)
+      throws IOException, AuthenticationException, InterruptedException {
     String appId = analyticJob.getAppId();
-    MapReduceApplicationData jobData = new MapReduceApplicationData();
     String jobId = Utils.getJobIdFromApplicationId(appId);
-    jobData.setAppId(appId).setJobId(jobId);
-    // Change job tracking url to job history page
     analyticJob.setTrackingUrl(_jhistoryWebAddr + jobId);
-    try {
 
+    MapReduceApplicationData jobData;
+    if (analyticJob.getJobStatus().equals("SUCCEEDED") || analyticJob.getJobStatus().equals("FAILED")) {
+      jobData = fetchCompletedJobsData(analyticJob);
+    } else if (analyticJob.getJobStatus().equals("RUNNING")) {
+      jobData = fetchRunningJobsData(analyticJob);
+    } else {
+      // Dr. Elephant doesn't support states other than SUCCEEDED, FAILED and UNDEFINED. Code shouldn't reach here.
```

Comment:
did you mean SUCCEEDED, FAILED and RUNNING ? Otherwise, your check is wrong as it's checking for RUNNING